### PR TITLE
Ui additions

### PR DESCRIPTION
**Pull request by heymansmile**
### Issue Link :link:
Additional functionality and fixes created while working on a tvOS project and while creating custom controls.

### Goals :soccer:
Add necessary functionality and fix bugs.

### Backwards-compatibility:
No breaking changes.

### CODE REVIEW

**Review by leontiy — DISMISSED** [review]

File: FueledUtils/UIExtensions.swift [hunk 1]
```diff
@@ -261,6 +261,22 @@ extension UIView {
...
+	/// - Parameters:
+	///   - opaque: Please refer to the parameters documentation for `UIGraphicsBeginImageContextWithOptions` for more info.
+	///   - scale: Please refer to the parameters documentation for `UIGraphicsBeginImageContextWithOptions` for more info.
+	/// - Returns: The image if it could be generated. If it couldn't, for example if the `UIView`'s width or height is 0, a crash will happen at runtime.
+	///
+	/// - Note: The returns is an implicitely unwrapped optional for backward-compatibility purpose, and will be made an optional in a future release (as well as not crash)
+	///
+	public func renderImage(opaque: Bool = false, scale: CGFloat = 0.0) -> UIImage {
+		return UIImage.draw(size: self.bounds.size, opaque: opaque, scale: scale) { context in
+			self.layer.render(in: context)
```
Comment (leontiy) [inline FueledUtils/UIExtensions.swift:277]: @Sissing Do you remember if this is the best way to do view rendering according to our ML/1001 research?

Comment (leontiy) [reply FueledUtils/UIExtensions.swift:277]: @heymansmile does this method render sublayers of the view?

Comment (heymansmile) [reply FueledUtils/UIExtensions.swift:277]: @leontiy yes sure it does everything.

Comment (stephanecopin) [reply FueledUtils/UIExtensions.swift:277]: I think the best way for `UIView` is still `self.drawHierarchy(in:, afterScreenUpdates:)`, as optionally you can just use the current render which can be much faster than re-rendering.
I think this method could make sense in a `CALayer` extension though.

File: FueledUtils/ButtonWithTitleAdjustment.swift [hunk 2]
```diff
@@ -88,12 +88,16 @@ open class ButtonWithTitleAdjustment: UIButton {
...
+			}
 			var attributes: [NSAttributedString.Key: Any] = [
-				NSAttributedString.Key.paragraphStyle: paragraphStyle,
-				NSAttributedString.Key.kern: self.adjustmentKerning,
+				.paragraphStyle: paragraphStyle,
+				.kern: self.adjustmentKerning
 			]
 			if let titleColor = self.titleColor(for: state) {
-				attributes[NSAttributedString.Key.foregroundColor] = titleColor
+				attributes[.foregroundColor] = titleColor
```
Comment (stephanecopin) [inline FueledUtils/ButtonWithTitleAdjustment.swift:100]: 👍

**Review by stephanecopin — COMMENTED** [review]
Just left a comment regarding the `UIView` extension for a refactoring

File: FueledUtils/UIExtensions.swift (hunk 1 above)
Comment (heymansmile) [reply FueledUtils/UIExtensions.swift:277]: @stephanecopin @leontiy this method can't be replaced with the existing one. The existing method works only for visible hierarchy and doesn't work when you, for example, create a custom control and need to apply an effect or gradient to a set of views and then make an image of their superposition in its initializer when it haven't been displayed yet.

Comment (leontiy) [reply FueledUtils/UIExtensions.swift:277]: @heymansmile Can you make it clear in the first line of the descriptions of both methods that one is for views in a presented view hierarchy and the other is for any view, regardless of the hierarchy status.

Comment (leontiy) [reply FueledUtils/UIExtensions.swift:277]: I also think `renderToImage` or `render(...) -> UIImage` would be clearer. The name of the function sounds like it accepts an image as a parameter when it doesn't.

**Review by leontiy — CHANGES_REQUESTED** [review]
1. can you please modify documentation comments to make the differences between snapshot and render methods crystal clear in the first line?
2. Let's consider clearer names for renderImage as that sounds like it takes an image which is a bit confusing.

**Comment by heymansmile** [comment]
@leontiy sure thing!

**Comment by leontiy** [comment]
@heymansmile sup here?

**Comment by heymansmile** [comment]
@leontiy I've renamed the method and amended the docs.

**Review by leontiy — APPROVED** [review]

**Comment by leontiy** [comment]
LGTM. @stephanecopin does this look good to you?

File: FueledUtils/ButtonWithTitleAdjustment.swift [hunk 3]
```diff
@@ -88,12 +88,16 @@ open class ButtonWithTitleAdjustment: UIButton {
 		let adjustedString = title.map { title -> NSAttributedString in
 			let paragraphStyle = NSMutableParagraphStyle()
 			paragraphStyle.lineSpacing = self.adjustmentLineSpacing
+			if let titleLabel = self.titleLabel {
+				paragraphStyle.lineBreakMode = titleLabel.lineBreakMode
+				paragraphStyle.alignment = titleLabel.textAlignment
```
Comment (stephanecopin) [inline FueledUtils/ButtonWithTitleAdjustment.swift:93]: 👍

**Review by stephanecopin — APPROVED** [review]
@leontiy looks good to me too, LGTM 👍

**Pull request merged**